File: Mega Sena/analise_mega.py
from collections import Counter
import re

def ler_resultados():
    numeros = []
    with open('resultados.txt', 'r', encoding='utf-8') as arquivo:
        for linha in arquivo:
            # Extrai apenas os números após o traço
            match = re.search(r'-\s*([\d,]+)', linha)
            if match:
                # Divide os números e adiciona à lista
                numeros_linha = match.group(1).split(',')
                numeros.extend(numeros_linha)
    return numeros

def analisar_numeros(arquivo, quantidade, tipo_ordenacao='mais'):
    """
    Analisa os números da Mega Sena a partir de um arquivo de resultados.
    
    Args:
        arquivo (str): Caminho do arquivo com os resultados
        quantidade (int): Quantidade de números a retornar
        tipo_ordenacao (str): 'mais' para mais sorteados, 'menos' para menos sorteados
    
    Returns:
        dict: Dicionário com os números e suas frequências
    """
    # Lista para armazenar todos os números sorteados
    todos_numeros = []
    
    try:
        # Abre o arquivo no modo de leitura
        with open(arquivo, 'r', encoding='utf-8') as f:
            # Itera sobre cada linha do arquivo
            for linha in f:
                # Remove espaços em branco e quebras de linha
                linha = linha.strip()
                # Verifica se a linha não está vazia
                if linha:
                    # Divide a linha em partes usando o caractere '|'
                    partes = linha.split('|')
                    # Verifica se a linha tem pelo menos 2 partes (data e números)
                    if len(partes) >= 2:
                        # Pega a parte dos números (segunda parte)
                        numeros_str = partes[1].strip()
                        # Divide a string dos números em uma lista
                        numeros = numeros_str.split()
                        # Converte cada número para inteiro e adiciona à lista
                        todos_numeros.extend([int(n) for n in numeros])
        
        # Verifica se foram encontrados números
        if not todos_numeros:
            return {'erro': 'Nenhum número encontrado no arquivo.'}
        
        # Conta a frequência de cada número usando Counter
        contagem = Counter(todos_numeros)
        
        # Ordena os números com base no tipo de ordenação
        if tipo_ordenacao == 'mais':
            # Ordena do mais frequente para o menos frequente
            numeros_ordenados = sorted(contagem.items(), key=lambda x: (-x[1], x[0]))
        else:
            # Ordena do menos frequente para o mais frequente
            numeros_ordenados = sorted(contagem.items(), key=lambda x: (x[1], x[0]))
        
        # Limita a quantidade de números retornados
        numeros_ordenados = numeros_ordenados[:quantidade]
        
        # Retorna um dicionário com os resultados
        return {
            'todos_numeros': numeros_ordenados,
            'tipo_ordenacao': tipo_ordenacao
        }
        
    except FileNotFoundError:
        # Retorna erro se o arquivo não for encontrado
        return {'erro': f'Arquivo {arquivo} não encontrado.'}
    except ValueError:
        # Retorna erro se houver problema na conversão dos números
        return {'erro': 'Erro ao processar os números no arquivo.'}
    except Exception as e:
        # Retorna erro genérico para outras exceções
        return {'erro': f'Erro ao processar o arquivo: {str(e)}'}

# Bloco principal que será executado quando o script for rodado diretamente
if __name__ == '__main__':
    # Define o caminho do arquivo de resultados
    arquivo = 'resultados.txt'
    
    try:
        # Solicita ao usuário a quantidade de números a mostrar
        quantidade = int(input('Quantos números você deseja ver? '))
        
        # Verifica se a quantidade é válida
        if quantidade < 1 or quantidade > 60:
            print('Por favor, insira um número entre 1 e 60.')
        else:
            # Realiza a análise dos números
            resultado = analisar_numeros(arquivo, quantidade)
            
            # Verifica se houve erro na análise
            if 'erro' in resultado:
                print(resultado['erro'])
            else:
                # Imprime os resultados formatados
                print(f'\nNúmeros mais sorteados:')
                for numero, frequencia in resultado['todos_numeros']:
                    print(f'Número {numero}: {frequencia} vezes')
                    
    except ValueError:
        # Trata erro se o usuário não inserir um número válido
        print('Por favor, insira um número válido.') 

File: Mega Sena/analise_mega.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_NUMEROS 60
#define MAX_LINHA 256
#define MAX_NUMEROS_LINHA 6

// Estrutura para armazenar número e sua frequência
typedef struct {
    int numero;
    int frequencia;
} NumeroFrequencia;

// Estrutura para armazenar contagem de pares e ímpares
typedef struct {
    int pares;
    int impares;
} ContagemParImpar;

// Função para comparar números (usada no qsort)
int compararNumeros(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

// Função para comparar frequências (usada no qsort)
int compararFrequencias(const void* a, const void* b) {
    NumeroFrequencia* nf1 = (NumeroFrequencia*)a;
    NumeroFrequencia* nf2 = (NumeroFrequencia*)b;
    
    if (nf1->frequencia != nf2->frequencia) {
        return nf2->frequencia - nf1->frequencia;
    }
    return nf1->numero - nf2->numero;
}

// Função para comparar frequências (ordem crescente)
int compararFrequenciasCrescente(const void* a, const void* b) {
    NumeroFrequencia* nf1 = (NumeroFrequencia*)a;
    NumeroFrequencia* nf2 = (NumeroFrequencia*)b;
    
    if (nf1->frequencia != nf2->frequencia) {
        return nf1->frequencia - nf2->frequencia;
    }
    return nf1->numero - nf2->numero;
}

// Função para extrair números de uma linha
int extrairNumeros(const char* linha, int* numeros) {
    int count = 0;
    char* token;
    char linha_copia[MAX_LINHA];
    char* inicio;
    
    strcpy(linha_copia, linha);
    
    // Procura pelo traço
    inicio = strchr(linha_copia, '-');
    if (!inicio) return 0;
    
    // Pula o traço e espaços
    inicio++;
    while (isspace(*inicio)) inicio++;
    
    // Divide a string pelos números
    token = strtok(inicio, ",");
    while (token != NULL && count < MAX_NUMEROS_LINHA) {
        // Remove espaços em branco
        while (isspace(*token)) token++;
        
        // Converte para número
        numeros[count++] = atoi(token);
        token = strtok(NULL, ",");
    }
    
    return count;
}

// Função para contar números pares e ímpares
ContagemParImpar contarParesImpares(NumeroFrequencia* frequencias, int total_frequencias, int quantidade) {
    ContagemParImpar contagem = {0, 0};
    int i;
    
    for (i = 0; i < quantidade && i < total_frequencias; i++) {
        if (frequencias[i].numero % 2 == 0) {
            contagem.pares++;
        } else {
            contagem.impares++;
        }
    }
    
    return contagem;
}

// Função principal de análise
void analisarNumeros(const char* arquivo, int quantidade, int tipo_ordenacao) {
    FILE* fp;
    char linha[MAX_LINHA];
    int numeros[MAX_NUMEROS * 1000];  // Array para todos os números
    int total_numeros = 0;
    NumeroFrequencia frequencias[MAX_NUMEROS];
    int total_frequencias = 0;
    int numeros_linha[MAX_NUMEROS_LINHA];
    int count;
    int i;
    int numero_atual;
    int freq_atual;
    ContagemParImpar contagem;
    
    // Abre o arquivo
    fp = fopen(arquivo, "r");
    if (fp == NULL) {
        printf("Erro ao abrir o arquivo %s\n", arquivo);
        return;
    }
    
    // Lê o arquivo linha por linha
    while (fgets(linha, sizeof(linha), fp)) {
        count = extrairNumeros(linha, numeros_linha);
        
        // Adiciona os números ao array principal
        for (i = 0; i < count; i++) {
            numeros[total_numeros++] = numeros_linha[i];
        }
    }
    
    fclose(fp);
    
    // Verifica se foram encontrados números
    if (total_numeros == 0) {
        printf("Nenhum número encontrado no arquivo.\n");
        return;
    }
    
    // Ordena os números
    qsort(numeros, total_numeros, sizeof(int), compararNumeros);
    
    // Calcula frequências
    numero_atual = numeros[0];
    freq_atual = 1;
    
    for (i = 1; i < total_numeros; i++) {
        if (numeros[i] == numero_atual) {
            freq_atual++;
        } else {
            frequencias[total_frequencias].numero = numero_atual;
            frequencias[total_frequencias].frequencia = freq_atual;
            total_frequencias++;
            
            numero_atual = numeros[i];
            freq_atual = 1;
        }
    }
    
    // Adiciona o último número
    frequencias[total_frequencias].numero = numero_atual;
    frequencias[total_frequencias].frequencia = freq_atual;
    total_frequencias++;
    
    // Ordena por frequência de acordo com o tipo escolhido
    if (tipo_ordenacao == 1) {
        qsort(frequencias, total_frequencias, sizeof(NumeroFrequencia), compararFrequencias);
        printf("\nNúmeros mais sorteados:\n");
    } else {
        qsort(frequencias, total_frequencias, sizeof(NumeroFrequencia), compararFrequenciasCrescente);
        printf("\nNúmeros menos sorteados:\n");
    }
    
    // Imprime os resultados
    for (i = 0; i < quantidade && i < total_frequencias; i++) {
        printf("Número %d: %d vezes\n", frequencias[i].numero, frequencias[i].frequencia);
    }
    
    // Conta e mostra a quantidade de números pares e ímpares
    contagem = contarParesImpares(frequencias, total_frequencias, quantidade);
    printf("\nQuantidade de números pares: %d\n", contagem.pares);
    printf("Quantidade de números ímpares: %d\n", contagem.impares);
}

void mostrarMenu() {
    printf("\n=== Menu de Análise da Mega Sena ===\n");
    printf("1. Números mais sorteados\n");
    printf("2. Números menos sorteados\n");
    printf("0. Sair\n");
    printf("Escolha uma opção: ");
}

int main() {
    char arquivo[] = "resultados.txt";
    int quantidade;
    int opcao;
    
    do {
        mostrarMenu();
        scanf("%d", &opcao);
        
        if (opcao == 0) {
            printf("Programa encerrado.\n");
            break;
        }
        
        if (opcao != 1 && opcao != 2) {
            printf("Opção inválida! Tente novamente.\n");
            continue;
        }
        
        printf("Quantos números você deseja ver? ");
        scanf("%d", &quantidade);
        
        if (quantidade < 1 || quantidade > 60) {
            printf("Por favor, insira um número entre 1 e 60.\n");
            continue;
        }
        
        analisarNumeros(arquivo, quantidade, opcao);
        printf("\nPressione Enter para continuar...");
        getchar(); // Limpa o buffer
        getchar();
        
    } while (opcao != 0);
    
    return 0;
} 